src/PosthogAnalytics.ts

```diff
@@ -0,0 +1,362 @@
+                autocapture: false,
+                mask_all_text: true,
+                mask_all_element_attributes: true,
+                // This only triggers on page load, which for our SPA isn't particularly useful.
+                // Plus, the .capture call originating from somewhere in posthog makes it hard
+                // to redact URLs, which requires async code.
+                //
+                // To raise this manually, just call .capture("$pageview") or posthog.capture_pageview.
+                capture_pageview: false,
+                sanitize_properties: this.sanitizeProperties,
+                respect_dnt: true,
+            });
+            this.enabled = true;
+        } else {
+            this.enabled = false;
+        }
+    }
+
+    private sanitizeProperties = (properties: posthog.Properties): posthog.Properties => {
+        // Callback from posthog to sanitize properties before sending them to the server.
+        //
+        // Here we sanitize posthog's built in properties which leak PII e.g. url reporting.
+        // See utils.js _.info.properties in posthog-js.
+
+        // Replace the $current_url with a redacted version.
+        // $redacted_current_url is injected by this class earlier in capture(), as its generation
+        // is async and can't be done in this non-async callback.
+        if (!properties['$redacted_current_url']) {
+            console.log("$redacted_current_url not set in sanitizeProperties, will drop $current_url entirely");
+        }
+        properties['$current_url'] = properties['$redacted_current_url'];
+        delete properties['$redacted_current_url'];
+
+        if (this.anonymity == Anonymity.Anonymous) {
+            // drop referrer information for anonymous users
+            properties['$referrer'] = null;
+            properties['$referring_domain'] = null;
+            properties['$initial_referrer'] = null;
+            properties['$initial_referring_domain'] = null;
+
+            // drop device ID, which is a UUID persisted in local storage
+            properties['$device_id'] = null;
+        }
+
+        return properties;
+    };
+
+    private static getAnonymityFromSettings(): Anonymity {
+        // determine the current anonymity level based on curernt user settings
+
+        // "Send anonymous usage data which helps us improve Element. This will use a cookie."
+        const analyticsOptIn = SettingsStore.getValue("analyticsOptIn", null, true);
+
+        // (proposed wording) "Send pseudonymous usage data which helps us improve Element. This will use a cookie."
+        //
+        // TODO: Currently, this is only a labs flag, for testing purposes.
+        const pseudonumousOptIn = SettingsStore.getValue("feature_pseudonymous_analytics_opt_in", null, true);
+
+        let anonymity;
+        if (pseudonumousOptIn) {
+            anonymity = Anonymity.Pseudonymous;
+        } else if (analyticsOptIn || analyticsOptIn === null) {
```

Comment:
Doesn't this mean the opt-in has become an opt-out...? Is that intentional?

Reply:
We want to be able to anonymously track onboarding to diagnose problems there, potentially prior to having asked the user for permission, and indeed prior to the user having interacted at all with our pages. For example:

* we want to be able to track immediate bounce rates from the user loading the welcome page and then just leaving
* we want to know if JS errors are occurring on the landing page or during onboarding
* we want to know aggregated load speeds for parts of our funnel

At present, we show the analytics opt-in dialogue on first login with that device.

So, yes, this test for `null` succinctly reflects that yes, we will be tracking anonymously when the user hasn't explicitly opted in or out.

One potential improvement would be to show opt-in very early in the onboarding. If we avoid tracking prior to that, we still won't be able to diagnose any problems occurring earlier than that, and it needs to be done in a way that won't hurt the funnel by adding an extra distraction. Also, ironically, we wouldn't be able to track the impact on the funnel from adding the opt-in to it.

This is a bit thorny, and the PR is still valuable without being able to track onboarding - so I'll take the null clause out for now and we can revisit later.

Reply:
Yeah, seems better to work through that more complex behaviour change in a separate, focused PR.